### PR TITLE
Vkitti

### PR DESCRIPTION
**Pull request by vict0rsch**
Add pretraining of the seg and depth heads on vkitti 2

## Data

* data is downloaded and extracted in `/miniscratch/_groups/ccai/data/vkitti2` (with an `about.txt` of course!)
* copied all exising jsons to `/miniscratch/_groups/ccai/data/jsons`
* put `train_kitti.json` and `val_kitti.json` there
* dataset is described there -> https://europe.naverlabs.com/research/computer-vision/proxy-virtual-worlds-vkitti-2/
* selected the entire `Scene02` as validation data
* Removed the `rain` `sunset` `fog` and `morning` variations

## Preprocessing

* added `data.py/process_kitti_seg()` function to parse and merge labels from the rgb png images of the segmentation maps
* depth is encoded as 1 px = 1cm so I just normalize the inverse depth in `tutils.py/get_normalized_depth_t()`
* input image is loaded as a regular `jpg` image

## Training

* I changed jsut a little the `trainer.py/setup()` interface with `self.all_loaders`, `self.kitty_display_images` and `self.base_display_images`
* Added a simple utility in the trainer `switch_data(self, to="kitti")` or `switch_data(self, to="base")` that changes `self.loaders` and `self.display_images` to point to either the kitti data or the base data  we are used to
* NOTHING changes in the rest of the code :) 
  * Except that we need to catch here and there potential mismatches between tasks and data keys (when pretraining with kitti, tasks are `msd` but in the pre-training phase there is no `m` data)

## Experiments

* I stopped [this one](https://www.comet.ml/vict0rsch/omnigan/d7b1eaab78c04c5da1ce1aa4c7c629dc?experiment-tab=chart&showOutliers=true&smoothing=0&transformY=smoothing&xAxis=step) but it shows the pre-training phase works
* [This one](https://www.comet.ml/vict0rsch/omnigan/e20bfefc76f84eacb01232f9f29042a2?experiment-tab=chart&showOutliers=true&smoothing=0&transformY=smoothing&xAxis=step) is running
* I added a test to check that everything works in `omnigan/test_trainer.py`

### CODE REVIEW

File: omnigan/data.py [hunk 1]
```diff
@@ -383,11 +498,20 @@ def check_samples(self):
 
 
 def get_loader(mode, domain, opts):
+    if (
+        domain != "kitti"
+        or not opts.train.kitti.pretrain
+        or not opts.train.kitti.batch_size
```
Comment (vict0rsch) [inline omnigan/data.py:504]: Because there is no discriminator we can have a much higher batch size (25)

File: omnigan/logger.py [hunk 2]
```diff
@@ -26,12 +26,14 @@ def log_comet_images(self, mode, domain):
         all_legends = ["Input"]
         task_legends = {}
 
+        if domain not in trainer.display_images[mode]:
```
Comment (vict0rsch) [inline omnigan/logger.py:29]: domains `r` and `rf` are not available during pre-training

File: omnigan/logger.py [hunk 3]
```diff
@@ -40,6 +42,9 @@ def log_comet_images(self, mode, domain):
                     if task == "p":
                         continue
 
+                    if task not in display_dict["data"]:
```
Comment (vict0rsch) [inline omnigan/logger.py:45]: `m` is not available at pre-training time

File: omnigan/trainer.py [hunk 4]
```diff
@@ -724,32 +739,70 @@ def setup(self, inference=False):
...
+
+            if self.kitti_pretrain:
+                self.kitty_display_images[mode] = {}
+            self.base_display_images[mode] = {}
+
             for domain, domain_loader in mode_dict.items():
-                dataset = self.loaders[mode][domain].dataset
+
+                if self.kitti_pretrain and domain == "kitti":
+                    target_dict = self.kitty_display_images
```
Comment (vict0rsch) [inline omnigan/trainer.py:751]: store images in the right `display` dict

File: omnigan/trainer.py [hunk 5]
```diff
@@ -766,11 +819,18 @@ def train(self):
...
 
             if (
-                self.logger.epoch == self.opts.gen.p.pl4m_epoch
+                self.logger.epoch == self.opts.gen.p.pl4m_epoch - 1
                 and get_num_params(self.G.painter) > 0
             ):
                 self.use_pl4m = True
 
+            if self.logger.epoch == self.opts.train.kitti.epochs - 1:
+                self.switch_data(to="base")
```
Comment (vict0rsch) [inline omnigan/trainer.py:828]: Pre-training will last for `opts.train.kitti.epochs` epochs

File: omnigan/trainer.py [hunk 6]
```diff
@@ -766,11 +819,18 @@ def train(self):
...
+                self.kitti_pretrain = False
+
+            if self.logger.epoch == self.opts.train.pseudo.epochs - 1:
```
Comment (vict0rsch) [inline omnigan/trainer.py:831]: Unify pseudo-labels training interface -> pseudo-signal is disabled after `opts.train.pseudo.epochs` epochs

File: omnigan/trainer.py [hunk 7]
```diff
@@ -837,7 +901,9 @@ def run_epoch(self):
             self.logger.global_step += 1
             self.logger.log_step_time(time())
 
-        self.update_learning_rates()
+        if not self.kitti_pretrain:
```
Comment (vict0rsch) [inline omnigan/trainer.py:904]: Don't update learning rates just yet

File: omnigan/tutils.py [hunk 8]
```diff
@@ -183,17 +183,23 @@ def normalize_tensor(t):
     return t
 
 
-def get_normalized_depth_t(arr, domain, normalize=False):
+def get_normalized_depth_t(tensor, domain, normalize=False):
     if domain == "r":
```
Comment (vict0rsch) [inline omnigan/tutils.py:187]: Rename `arr` to tensor as it is misleading on the data  type (+ needs docstrings!)

File: shared/trainer/defaults.yaml [hunk 9]
```diff
@@ -1,4 +1,4 @@
-output_path: /network/tmp1/schmidtv/yb_runs/test_v1
+output_path: /miniscratch/_groups/ccai/trash
```
Comment (vict0rsch) [inline shared/trainer/defaults.yaml:1]: ⚠️ Do not rely on this directory as it may be emptied at any point in time

File: shared/trainer/defaults.yaml [hunk 10]
```diff
@@ -24,15 +24,17 @@ tasks: [d, s, m, p] # [p] [m, s, d]
 data:
   max_samples: -1 # -1 for all, otherwise set to an int to crop the training data size
   files: # if one is not none it will override the dirs location
-    base: /network/tmp1/ccai/data/omnigan/base/
+    base: /miniscratch/_groups/ccai/data/jsons
```
Comment (vict0rsch) [inline shared/trainer/defaults.yaml:27]: New json location

File: test_trainer.py [hunk 11]
```diff
@@ -191,45 +202,52 @@ def delete_on_exit(exp):
...
         },
         {
-            "__doc": "MSDP with End-to-end",
+            "__doc": "MSDP with End-to-end",  # 8
             "__pl4m": True,
             "domains": ["rf", "r", "s"],
             "tasks": ["m", "s", "d", "p"],
         },
+        {
+            "__doc": "Kitti pretrain",  # 9
```
Comment (vict0rsch) [inline test_trainer.py:245]: The kitti test

File: test_trainer.py [hunk 12]
```diff
@@ -140,8 +140,19 @@ def delete_on_exit(exp):
     parser = ArgumentParser()
     parser.add_argument("--no_delete", action="store_true", default=False)
     parser.add_argument("--no_end_to_end", action="store_true", default=False)
+    parser.add_argument("--include", "-i", nargs="+", default=[])
+    parser.add_argument("--exclude", "-e", nargs="+", default=[])
     args = parser.parse_args()
```
Comment (vict0rsch) [inline test_trainer.py:145]: add `include` and `exclude` args to select tests to run

To run the kitti test only:

```
python test_trainer.py --no_delete -i 9
```

File: utils_scripts/make_vkitti2_jsons.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
+from pathlib import Path
```
Comment (vict0rsch) [inline utils_scripts/make_vkitti2_jsons.py:1]: Script to create the kitti jsons
